crates/ra_editor/src/line_index_utils.rs

```diff
@@ -0,0 +1,396 @@
+                    NextNewlines::Use
+                } else if step_pos <= edit.delete.end() {
+                    let iter = OffsetNewlineIter {
+                        offset: edit.delete.start(),
+                        text: &edit.insert,
+                    };
+                    // empty slice
+                    edit.insert = &edit.insert[edit.insert.len()..];
+                    NextNewlines::ReplaceMany(iter)
+                } else {
+                    let iter = OffsetNewlineIter {
+                        offset: edit.delete.start(),
+                        text: &edit.insert,
+                    };
+                    // empty slice
+                    edit.insert = &edit.insert[edit.insert.len()..];
+                    self.advance_edit();
+                    NextNewlines::AddMany(iter)
+                }
+            }
+            None => NextNewlines::Use,
+        };
+        res
+    }
+
+    fn translate_range(&self, range: TextRange) -> TextRange {
+        if self.acc_diff == 0 {
+            range
+        } else {
+            let start = self.translate(range.start());
+            let end = self.translate(range.end());
+            TextRange::from_to(start, end)
+        }
+    }
+
+    fn translate(&self, x: TextUnit) -> TextUnit {
+        if self.acc_diff == 0 {
+            x
+        } else {
+            TextUnit::from((x.to_usize() as i64 + self.acc_diff) as u32)
+        }
+    }
+
+    fn translate_step(&self, x: &Step) -> Step {
+        if self.acc_diff == 0 {
+            x.clone()
+        } else {
+            match x {
+                &Step::Newline(n) => Step::Newline(self.translate(n)),
+                &Step::Utf16Char(r) => Step::Utf16Char(self.translate_range(r)),
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct RunningLineCol {
+    line: u32,
+    last_newline: TextUnit,
+    col_adjust: TextUnit,
+}
+
+impl RunningLineCol {
+    fn new() -> RunningLineCol {
+        RunningLineCol {
+            line: 0,
+            last_newline: TextUnit::from(0),
+            col_adjust: TextUnit::from(0),
+        }
+    }
+
+    fn to_line_col(&self, offset: TextUnit) -> LineCol {
+        LineCol {
+            line: self.line,
+            col_utf16: ((offset - self.last_newline) - self.col_adjust).into(),
+        }
+    }
+
+    fn add_line(&mut self, newline: TextUnit) {
+        self.line += 1;
+        self.last_newline = newline;
+        self.col_adjust = TextUnit::from(0);
+    }
+
+    fn adjust_col(&mut self, range: &TextRange) {
+        self.col_adjust += range.len() - TextUnit::from(1);
+    }
+}
+
+pub fn translate_offset_with_edit(
+    line_index: &LineIndex,
+    offset: TextUnit,
+    edits: &[AtomTextEdit],
+) -> LineCol {
+    let mut sorted_edits: Vec<&AtomTextEdit> = Vec::with_capacity(edits.len());
+    for edit in edits {
+        let insert_index =
+            sorted_edits.upper_bound_by_key(&edit.delete.start(), |x| x.delete.start());
```

Comment:
The `TextEdit` struct grantees that its AtomicEdits are disjoint and sorted. Can we accept an `TextEdit` here instead of the slice and avoid the sorting step?

Reply:
Yes that makes sense. Although reviewing the code, when `translate_offset_with_edit` is called we don't have a `TextEdit`. Here https://github.com/vemoo/rust-analyzer/blob/0156a538089828340a823ed02da8970bf4f1175b/crates/ra_lsp_server/src/conv.rs#L204 the edits come from `SourceFileNodeEdit` https://github.com/vemoo/rust-analyzer/blob/0156a538089828340a823ed02da8970bf4f1175b/crates/ra_analysis/src/lib.rs#L124

Should we create a new `TextEdit` before calling `translate_offset_with_edit` or should `SourceFileNodeEdit` have a `TextEdit`?

Reply:
So if you rebase this PR onto the current master it should be ready to go?

Reply:
Ok, I will do that later today

Reply:
I've rebased and changed `translate_offset_with_edit` to accept `TextEdit`.